app/src/main/java/de/westnordost/streetcomplete/quests/smoking/AddSmoking.kt

```diff
@@ -0,0 +1,57 @@
+package de.westnordost.streetcomplete.quests.smoking
+
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.mapdata.filter
+import de.westnordost.streetcomplete.data.meta.updateWithCheckDate
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
+import de.westnordost.streetcomplete.data.osm.osmquests.Tags
+import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.CITIZEN
+import de.westnordost.streetcomplete.ktx.arrayOfNotNull
+
+class AddSmoking : OsmFilterQuestType<SmokingAllowed>() {
+
+    private val elementFilterBasic = """
+        nodes, ways with
+        (
+          amenity ~ bar|cafe|pub|biergarten|restaurant|nightclub|stripclub
+          or leisure ~ outdoor_seating
+        )
+    """
+
+    override val elementFilter = elementFilterBasic + """
+	and (!smoking or smoking older today -8 years)
+    """
+
+    override val changesetComment = "Add smoking status"
+    override val wikiLink = "Key:smoking"
+    override val icon = R.drawable.ic_quest_smoke
+    override val isReplaceShopEnabled = true
+    override val defaultDisabledMessage = R.string.default_disabled_msg_go_inside_regional_warning
+
+    override val questTypeAchievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) =
+        if (hasFeatureName(tags))
+            R.string.quest_smoking_name_type_title
+        else
+            R.string.quest_smoking_no_name_title
+
+    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> {
+        val name = tags["name"] ?: tags["brand"]
+        return arrayOfNotNull(name, featureName.value)
+    }
+
+    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
+        getMapData().filter(elementFilterBasic)
+
+    override fun createForm() = SmokingAllowedAnswerForm()
+
+    override fun applyAnswerTo(answer: SmokingAllowed, tags: Tags, timestampEdited: Long) {
+	tags.updateWithCheckDate("smoking", answer.osmValue)
+    }
+
+    private fun hasFeatureName(tags: Map<String, String>): Boolean =
```

Comment:
See https://github.com/streetcomplete/StreetComplete/blob/12e95b923acc47b29a2097a15c25644c5bd83f22/app/src/main/java/de/westnordost/streetcomplete/quests/opening_hours_signed/CheckOpeningHoursSigned.kt#L115-L120

AFAIK `hasFeatureName` is consistently used "featureDictionaryFuture preset has name for this set of tags" and `hasProperName` is used for "has name tags such as `name`/`brand`/`operator` (which count as naming tags depends on an object!)"

I would change name here to `hasProperName`

Maybe add `hasFeatureName` that would be used for extra check ensuring that things returned by filter all have names in presets, but I am not entirely sure here is it needed.

I need to check my explanation in #3648 is being confused/misleading

Reply:
done in 3b75cf814a62eebd3eb2893d66bdb5773fd1031c